# **Actividad: Ajuste de redes neuronales**

Héctor Manuel Cárdenas Yáñez - A01634615

In [ ]:
# Importar Librerías

import pandas as pd
import numpy as np

from sklearn.neural_network import MLPRegressor, MLPClassifier
from sklearn.model_selection import KFold, GridSearchCV, cross_val_predict
from sklearn.metrics import mean_squared_error, mean_absolute_error, classification_report

## Ejercicio 1

In [ ]:
# Cargar los datos
df = pd.read_csv('/content/crime_data.csv')
df.head()

State    VR    MR     M     W     H     P     S
0    AK   761   9.0  41.8  75.2  86.6   9.1  14.3
1    AL   780  11.6  67.4  73.5  66.9  17.4  11.5
2    AR   593  10.2  44.7  82.9  66.3  20.0  10.7
3    AZ   715   8.6  84.7  88.6  78.7  15.4  12.1
4    CA  1078  13.1  96.7  79.3  76.2  18.2  12.5

In [ ]:
# Eliminamos columnas innecesarias
eliminar = ['VR', 'P']
df = df.drop(eliminar, axis=1)

In [ ]:
# Definimos las variables
x = df[['M', 'W', 'H', 'S']].values # Variables predictoras
y = df['MR'].values # Variable de respuesta

### 1. Evalúa con validación cruzada un modelo pereceptrón multicapa para las variables que se te asignaron para este ejercicio.

In [ ]:
# Features
features = x
n_features = len(x)

# Fit regression model using the complete dataset and calculate MSE and MAE
regr = MLPRegressor(hidden_layer_sizes=(20, 20), max_iter=20000)
regr.fit(x, y)

y_pred = regr.predict(x)
print('MSE: ', mean_squared_error(y, y_pred))
print("MAE: ", mean_absolute_error(y, y_pred))

# Evaluate regression model using k-fold cross-validation
n_folds = 5
kf = KFold(n_splits=n_folds, shuffle = True)

mse_cv = []
mae_cv = []
for train_index, test_index in kf.split(x):
    # Training phase
    x_train = x[train_index, :]
    y_train = y[train_index]
    regr_cv = MLPRegressor(hidden_layer_sizes=(20, 20), max_iter=20000)
    regr_cv.fit(x_train, y_train)

    # Test phase
    x_test = x[test_index, :]
    y_test = y[test_index]
    y_pred = regr_cv.predict(x_test)

    # Calculate MSE and MAE
    mse_i = mean_squared_error(y_test, y_pred)
    print('mse = ', mse_i)
    mse_cv.append(mse_i)

    mae_i = mean_absolute_error(y_test, y_pred)
    print('mae = ', mae_i)
    mae_cv.append(mae_i)

print('MSE:', np.average(mse_cv), '  MAE:', np.average(mae_cv))

MSE:  2.3263091269118688
MAE:  1.1990011073294324
mse =  27.28026165787329
mae =  4.288196680704442
mse =  569.7720151479743
mae =  12.446261082114061
mse =  8.115622033211768
mae =  2.4287489425956235
mse =  161.3614751222328
mae =  5.841685578382941
mse =  5.140627821321422
mae =  1.921992735126101
MSE: 154.33400035652275   MAE: 5.385377003784633


### 2. Agrega al conjunto de datos columnas que representen los cuadrados de las variables predictoras (por ejemplo, M2, W2), así como los productos entre pares de variables (por ejemplo, PxS, MxW). Evalúa un modelo perceptrón multicapa para este nuevo conjunto de datos.

In [ ]:
# Definimos los cuadrados y productos de las variables predictoras
df['M2'] = df['M'] ** 2
df['W2'] = df['W'] ** 2
df['H2'] = df['H'] ** 2
df['S2'] = df['S'] ** 2

df['MxW'] = df['M'] * df['W']
df['MxH'] = df['M'] * df['H']
df['MxS'] = df['M'] * df['S']

df['WxH'] = df['W'] * df['H']
df['WxS'] = df['W'] * df['S']

df['HxS'] = df['H'] * df['S']

# Redefinimos los valores "x" y "y"
x = df[['M', 'W', 'H', 'S','M2', 'W2', 'H2', 'S2', 'MxW', 'MxH', 'MxS', 'WxH', 'WxS', 'HxS']].values
y = df['MR'].values

In [ ]:
# Features
features = x
n_features = len(x)

# Fit regression model using the complete dataset and calculate MSE and MAE
regr = MLPRegressor(hidden_layer_sizes=(20, 20), max_iter=20000)
regr.fit(x, y)

y_pred = regr.predict(x)
print('MSE: ', mean_squared_error(y, y_pred))
print("MAE: ", mean_absolute_error(y, y_pred))

# Evaluate regression model using k-fold cross-validation
n_folds = 5
kf = KFold(n_splits=n_folds, shuffle = True)

mse_cv = []
mae_cv = []
for train_index, test_index in kf.split(x):
    # Training phase
    x_train = x[train_index, :]
    y_train = y[train_index]
    regr_cv = MLPRegressor(hidden_layer_sizes=(20, 20), max_iter=20000)
    regr_cv.fit(x_train, y_train)

    # Test phase
    x_test = x[test_index, :]
    y_test = y[test_index]
    y_pred = regr_cv.predict(x_test)

    # Calculate MSE and MAE
    mse_i = mean_squared_error(y_test, y_pred)
    print('mse = ', mse_i)
    mse_cv.append(mse_i)

    mae_i = mean_absolute_error(y_test, y_pred)
    print('mae = ', mae_i)
    mae_cv.append(mae_i)

print('MSE:', np.average(mse_cv), '  MAE:', np.average(mae_cv))

MSE:  9.027981148582542
MAE:  2.3678548029262627
mse =  181319.9892995284
mae =  359.67854171809006
mse =  45529.60676069503
mae =  186.0857461838582
mse =  50982.20081256821
mae =  195.7689785566457
mse =  713.1068114324815
mae =  18.56224333255103
mse =  82.90743363693362
mae =  5.866265917414897
MSE: 55725.562223572204   MAE: 153.19235514171197


### 3. Viendo los resultados de regresión, desarrolla una conclusión sobre los siguientes puntos:

**A. ¿Consideras que el modelo perceptrón multicapa es efectivo para modelar los datos del problema? ¿Por qué?**

Analizando los resultados de ambos métodos podemos darnos cuenta que el modelo sin inclusión de cuadrados presenta un mejor ahuste general de MSE: 154.33400035652275   MAE: 5.385377003784633, a comparación de los resultaods incluyendo los cuadrados de MSE: 55725.562223572204   MAE: 153.19235514171197. Sin embargo, cabe resaltar que ningunos de los modelos presenta una solución eficiente, ambos tienen errores bastante altos, por lo que no evaluan efectivamente los datos.

**B. ¿Qué modelo es mejor para los datos de criminalidad, el lineal o el perceptrón multicapa? ¿Por qué?**

En base a las métricas de rendimiento, el modelo de regresión lineal es preferible sobre el perceptrón multicapa para los datos de criminalidad. Los resultados de MSE y MAE son
considerablemente bajos para indicar que el modelo de perceptrón multicapa sea más eficiente que modelo de regresión lineal.


## Ejercicio 2

In [ ]:
# Cargar los datos
data_2 = np.loadtxt('/content/M_1.txt')

# Definimos las variables regresoras y dependiente
x = data_2[:,2:]
y = data_2[:,0]

### 1. Evalúa un modelo perceptrón multicapa con validación cruzada utilizando al menos 5 capas de 20 neuronas.

In [ ]:
%%time
features = x
n_features = len(x)

# Train model with all observations (two layers of 100 neurons)
clf = MLPClassifier(hidden_layer_sizes=(20, 20, 20, 20, 20), max_iter=10000)
clf.fit(x, y)

# Evaluate model (two layers of 100 neurons)
y_pred = cross_val_predict(MLPClassifier(hidden_layer_sizes=(20, 20, 20, 20, 20), max_iter=10000), x, y)
print(classification_report(y, y_pred))

              precision    recall  f1-score   support

         1.0       0.95      1.00      0.97        90
         2.0       0.91      0.89      0.90        90
         3.0       0.92      0.92      0.92        90
         4.0       1.00      0.98      0.99        90
         5.0       0.97      0.94      0.96        90
         6.0       0.85      0.86      0.85        90
         7.0       0.99      0.99      0.99        90

    accuracy                           0.94       630
   macro avg       0.94      0.94      0.94       630
weighted avg       0.94      0.94      0.94       630

CPU times: user 5.57 s, sys: 2.9 s, total: 8.47 s
Wall time: 5.17 s


### 2. Evalúa un modelo perceptrón multicapa con validación cruzada, pero encontrando el número óptimo de capas y neuronas de la red.

In [ ]:
%%time
# Optimal number of layers and neurons
num_layers = np.arange(1, 20, 5)
num_neurons = np.arange(10, 110, 20)

layers = []

for l in num_layers:
    for n in num_neurons:
        layers.append(l*[n])

clf = GridSearchCV(MLPClassifier(max_iter=10000), {'hidden_layer_sizes': layers}, cv = 5)
clf.fit(x, y)
print(clf.best_estimator_)

MLPClassifier(hidden_layer_sizes=[50], max_iter=10000)
CPU times: user 2min 53s, sys: 1min 10s, total: 4min 4s
Wall time: 2min 17s


### 3. Prepara el modelo perceptrón multicapa:

A. Opten los hiperparámetros óptimos de capas y neuronas de la red.

B. Con los hiperparámetros óptimos, ajusta el modelo con todos los datos.

In [ ]:
%%time
# Evaluate model finding the optimal number of layers and neurons
clf = GridSearchCV(MLPClassifier(max_iter=10000), {'hidden_layer_sizes': layers}, cv = 5)
y_pred = cross_val_predict(clf, x, y, cv = 5)
print(classification_report(y, y_pred))

              precision    recall  f1-score   support

         1.0       0.97      0.99      0.98        90
         2.0       0.94      0.98      0.96        90
         3.0       0.97      0.96      0.96        90
         4.0       1.00      0.98      0.99        90
         5.0       0.98      0.98      0.98        90
         6.0       0.93      0.91      0.92        90
         7.0       0.99      0.98      0.98        90

    accuracy                           0.97       630
   macro avg       0.97      0.97      0.97       630
weighted avg       0.97      0.97      0.97       630

CPU times: user 10min 36s, sys: 4min 28s, total: 15min 5s
Wall time: 8min 22s


### 4. Contesta lo siguientes:
A. ¿Observas alguna mejora importante al optimizar el tamaño de la red? ¿Es el resultado que esperabas? Argumenta tu respuesta.

El modelo inicial tiene un excelente rendimiento de un 94% de presición. Implementando las capas de redes neuronales el modelos mejoró a un promedio de 97%, por lo que en general mejoró un 0.03%.

B. ¿Qué inconvenientes hay al encontrar el tamaño óptimo de la red? ¿Por qué?

Realmente el único inconveniente que hay es el tiempo y la capacidad. El tiempo es basicamente el doble o más. Considerando la mejora de un 0.03% en general se debería analizar si es más óptimo implemetar las capas o no. Sin embargo la mejora si es evidente.

In [ ]:
# Guardar en Drive como PDF
!apt-get install -y texlive-xetex

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
!jupyter nbconvert --to pdf '/content/drive/MyDrive/Colab Notebooks/RedesNeuronales.ipynb' --output-dir='/content/drive/MyDrive/Colab Notebooks'

[NbConvertApp] Converting notebook /content/drive/MyDrive/Colab Notebooks/RedesNeuronales.ipynb to pdf
[NbConvertApp] Writing 48281 bytes to notebook.tex
[NbConvertApp] Building PDF
[NbConvertApp] Running xelatex 3 times: ['xelatex', 'notebook.tex', '-quiet']
[NbConvertApp] Running bibtex 1 time: ['bibtex', 'notebook']
[NbConvertApp] WARNING | bibtex had problems, most likely because there were no citations
[NbConvertApp] PDF successfully created
[NbConvertApp] Writing 61844 bytes to /content/drive/MyDrive/Colab Notebooks/RedesNeuronales.pdf
